src/endpoint/s3/s3_rest.js

```diff
@@ -277,23 +277,27 @@ async function authorize_request_policy(req) {
         if (is_owner || is_iam_account_and_same_root_account_owner) return;
         throw new S3Error(S3Error.AccessDenied);
     }
-    let permission;
+    // in case we have bucket policy
+    let permission_by_id;
+    let permission_by_name;
     // In NC, we allow principal to be:
     // 1. account name (for backwards compatibility)
     // 2. account id
     // we start the permission check on account identifier intentionally
     if (account_identifier_id) {
-        permission = await s3_bucket_policy_utils.has_bucket_policy_permission(
+        permission_by_id = await s3_bucket_policy_utils.has_bucket_policy_permission(
             s3_policy, account_identifier_id, method, arn_path, req);
+        dbg.log3('authorize_request_policy: permission_by_id', permission_by_id);
     }
 
-    if ((!account_identifier_id || permission === "IMPLICIT_DENY") && account.owner === undefined) {
-        permission = await s3_bucket_policy_utils.has_bucket_policy_permission(
+    if ((!account_identifier_id || permission_by_id !== "DENY") && account.owner === undefined) {
```

Comment:
Somewhat nitpicking, but you can remove the 
`|| permission_by_id !== 'DENY'`
because if permission_by_id === 'DENY' the result would be S3Error.AccessDenied anyway.

If you're concerned about efficiency then it would be more efficient to throw the S3Error.AccessDenied in line 292.

Reply:
I changed the code to your suggestion.